### PR TITLE
MOBT-84: Cell method updates for precipitation phases

### PR DESCRIPTION
**Pull request by bayliffe**
Accumulation diagnostic cubes possess a cell method that highlights over which dimension (e.g. time, space) the diagnostic has been accumulated. This cell method may further specify what has been accumulated, where this is important for probabilistic cubes to make clear that it is not the probabilities that have been accumulated but instead the underlying physical diagnostic.

When we apply phase information to precipitation accumulations at present the cell method remains unchanged, e.g.
```
probability_of_lwe_thickness_of_snowfall_amount_above_threshold
     ...
     Cell methods:
          sum: time (of thickness_of_precipitation_amount)
```
The cell method and diagnostic name are thus mismatched. This PR addresses this issue by ensuring that when cubes are combined by multiplication and a new diagnostic name is provided, any cell method that makes reference to the diagnostic name is updated. The above would therefore become:
```
probability_of_lwe_thickness_of_snowfall_amount_above_threshold
     ...
     Cell methods:
          sum: time (of lwe_thickness_of_snowfall_amount)
```

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

File: improver_tests/acceptance/SHA256SUMS [hunk 1]
```diff
@@ -79,6 +79,9 @@ f3848563a1ff89a06e487036389c49849d345b7d1cab8e6ac3a961ecf9c879c6  ./combine/boun
 e88c3624b3b290db8c8203ca8802f2d8f341da0879bebaec5bb4570aa54bc27c  ./combine/broadcast/input_larger_cube.nc
 b5ca2030a23ba6440c712a659f6f4f3dcbe8566afeef1a68e2e5a6a9b6e7484e  ./combine/broadcast/input_smaller_cube.nc
 02a9f7a2fe3e80d9e7d43e87e7c8031f6e82be2137a02cd9cfe8c93f79906a39  ./combine/broadcast/kgo.nc
+e7a0d4e2c2541dfea51a71d0689ba100290de8986d79f5d858101f26c77cd741  ./combine/cellmethods/kgo.nc
```
Comment (gavinevans) [inline improver_tests/acceptance/SHA256SUMS:82]: These files contain some unexpected attributes i.e. `NCO` and `history`, which cause warnings in the interpret_metadata CLI.

**Review by gavinevans — CHANGES_REQUESTED** [review]
Thanks for these modifications, @bayliffe 👍 

I've added a few comments below.

File: improver_tests/cube_combiner/test_CubeMultiplier.py [hunk 2]
```diff
@@ -144,6 +160,37 @@ def test_exception_for_single_entry_cubelist(self):
...
+
+        result = CubeMultiplier()(cubelist, new_cube_name, broadcast_to_threshold=True)
+        self.assertEqual(result.cell_methods[0], expected)
+
+    def test_unmodified_cell_methods(self):
+        """Test that plugin leaves cell methods that are diagnostic name
+        agnostic unchanged."""
+
+        cell_methods = list(self.cube.cell_methods)
+        additional_cell_method = CellMethod("sum", coords="longitude")
```
Comment (MoseleyS) [inline improver_tests/cube_combiner/test_CubeMultiplier.py:179]: If I change this as below, the test fails.
```suggestion
        additional_cell_method = CellMethod("sum", coords="longitude", comments="Does not match cube name")
```

**Comment by bayliffe** [comment]
@MoseleyS regarding the in-vicinity diagnostics, please read the comment on the original ticket: https://github.com/metoppv/mo-blue-team/issues/84#issuecomment-929899771

Bruce decided that these should not be modified to include the in-vicinity diagnostic. We therefore need to modify the metadata-interpreter rather than any functional code, so this is less urgent. I will note the need for a related ticket as an additional acceptance criteria.

File: improver/cube_combiner.py [hunk 3]
```diff
@@ -285,6 +286,55 @@ def _coords_are_broadcastable(coord1: DimCoord, coord2: DimCoord) -> bool:
...
+
+        Returns:
+            A list of modified cell methods to replace the originals.
+        """
+        try:
+            diagnostic_name = get_threshold_coord_name_from_probability_name(
+                diagnostic_name
+            )
+        except ValueError:
+            pass
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/cube_combiner.py:318]: I had missed that the multiplier specified it was only for use with probabilities. That makes this simpler, so I've removed the try/except.

File: improver/cube_combiner.py [hunk 4]
```diff
@@ -326,22 +376,32 @@ def process(
 
         result = self._combine_cube_data(cube_list)
 
+        # Used for renaming the threshold coordinate and modifying cell methods
+        # where neccessary; excludes the in_vicinity component.
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/cube_combiner.py:380]: You're just jealous of my avant-garde approach to spelling. In years to come all letters will be paired, but for now I will capitulate to your convention.

File: improver_tests/acceptance/SHA256SUMS (hunk 1 above)
Comment (bayliffe) [reply improver_tests/acceptance/SHA256SUMS:82]: Thanks for picking that up. I hadn't appreciated that NCO tools add history attributes when working on the netcdf files, though it makes sense. I've removed the extraneous attributes from the input accumulation and regenerated the KGO.

File: improver/cube_combiner.py [hunk 5]
```diff
@@ -285,6 +286,55 @@ def _coords_are_broadcastable(coord1: DimCoord, coord2: DimCoord) -> bool:
             (len(coord1.points) == 1) or (len(coord2.points) == 1)
         )
 
+    @staticmethod
+    def _update_cell_methods(
+        cell_methods: List, diagnostic_name: str, new_diagnostic_name: str
+    ) -> List:
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/cube_combiner.py:292]: Changed.

File: improver/cube_combiner.py [hunk 6]
```diff
@@ -285,6 +286,55 @@ def _coords_are_broadcastable(coord1: DimCoord, coord2: DimCoord) -> bool:
...
+            pass
+
+        new_cell_methods = []
+        for cell_method in cell_methods:
+            try:
+                (cell_comment,) = cell_method.comments
+            except ValueError:
+                new_cell_methods.append(cell_method)
+            else:
+                if diagnostic_name in cell_comment:
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/cube_combiner.py:327]: We absolutely do. Added. I've also added a commented cell method to the test you highlighted below, but a comment that does not contain the comment name, to test this else block.

File: improver_tests/cube_combiner/test_CubeMultiplier.py (hunk 2 above)
Comment (bayliffe) [reply improver_tests/cube_combiner/test_CubeMultiplier.py:179]: Issue fixed and test extended as written above.

File: improver/cube_combiner.py [hunk 7]
```diff
@@ -285,6 +286,55 @@ def _coords_are_broadcastable(coord1: DimCoord, coord2: DimCoord) -> bool:
...
+        Args:
+            cell_methods:
+                The cell methods found on the cube that is being used as the
+                metadata template.
+            diagnostic_name:
+                The full name of the metadata template cube. This will be
+                stripped of any probability and vicinity components to provide
+                the base name to look for in the cell methods.
+            new_diagnostic_name:
+                The new diagnostic name to use in the modified cell methods.
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/cube_combiner.py:308]: I've modified this to use the variable `probabilistic_name` in the interface. This then gets stripped of the vicinity and probability bits to become `diagnostic_name`, making that and `new_diagnostic_name` the same in form. Hopefully that's clearer.

**Review by MoseleyS — APPROVED** [review]

**Review by gavinevans — APPROVED** [review]

**Pull request merged**